src/event-scope.ts

```diff
@@ -1,177 +1,191 @@
+
+						// Delete the field to indicate that raising the exit event suceeded
+						delete this._exitEventVersion;
 					}
-
-					// Move subscribers to the parent scope
-					exitSubscriptions.forEach(sub => {
-						if (!sub.isOnce || !sub.isExecuted) {
-							this.parent.onExit.subscribe(sub.handler);
+					else {
+						try {
+							// Attempt to move subscribers to the parent scope
+							this.parent.receiveEventHandlers(exitSubscriptions);
+						}
+						catch (e) {
+							this.dispose({ abort: true });
+							throw e;
 						}
-					});
-
-					if (this.parent._exitEventVersion !== undefined) {
-						this.parent._exitEventVersion++;
 					}
 				}
-
-				// Clear the events to ensure that they aren't
-				// inadvertantly raised again through this scope
-				this.onAbort.clear();
-				this.onExit.clear();
 			}
+
+			// Clear the events to ensure that they aren't inadvertantly raised again through this scope
+			this._onExit.clear();
 		}
 		finally {
 			// The event scope is no longer active
 			this.isActive = false;
-
-			if (EventScope$current && EventScope$current === this) {
-				// Roll back to the closest active scope
-				while (EventScope$current && !EventScope$current.isActive) {
-					EventScope$current = EventScope$current.parent;
-				}
-			}
-		}
-	}
-}
-
-export function EventScope$onExit(callback: Function, thisPtr: any = null): void {
-	if (EventScope$current === null) {
-		// Immediately invoke the callback
-		if (thisPtr) {
-			callback.call(thisPtr);
-		}
-		else {
-			callback();
 		}
 	}
-	else if (!EventScope$current.isActive) {
-		throw new Error("The current event scope cannot be inactive.");
-	}
-	else {
-		// Subscribe to the exit event
-		EventScope$current.onExit.subscribe(callback.bind(thisPtr));
-	}
-}
 
-export function EventScope$onAbort(callback: Function, thisPtr: any = null): void {
-	if (EventScope$current !== null) {
-		if (!EventScope$current.isActive) {
-			throw new Error("The current event scope cannot be inactive.");
+	private receiveEventHandlers(subscriptions: EventSubscription<EventScope, EventScopeExitEventArgs>[]) {
+		var maxNesting = this.settings.maxExitingTransferCount - 1;
+		if (this._exitEventVersion >= maxNesting) {
+			throw new Error("Exceeded max scope event transfer.");
 		}
 
-		// Subscribe to the abort event
-		EventScope$current.onAbort.subscribe(callback.bind(thisPtr));
-	}
-}
+		// Move subscribers to the parent scope
+		subscriptions.forEach(sub => {
+			if (!sub.isOnce || !sub.isExecuted) {
+				this._onExit.subscribe(sub.handler);
+			}
+		});
 
-export function EventScope$perform(callback: Function, thisPtr: any = null): void {
-	// Create an event scope
-	var scope = new EventScope();
-	try {
-		// Invoke the callback
-		if (thisPtr) {
-			callback.call(thisPtr);
-		}
-		else {
-			callback();
+		if (this._exitEventVersion !== undefined) {
+			this._exitEventVersion++;
```

Comment:
Just want to make sure I'm understanding this correctly: `_exitEventVersion` only grows if we keep transferring events onto the same parent scope?

Reply:
Hm... I'm honestly not sure about this logic. I'll take another look at it.

Reply:
Yeah, that's the gist of it. I got hung up on the fact that the parent is exiting, but what happens is the list of exit event subscriptions keeps growing as the subscribers are invoked. For example, exiting a scope (ex: calculation of property A) triggers calculation of property B which exits and transfers its events to the parent, which includes a callback to trigger calculation of property C, which does the same thing for property B (or A), and then back to C again, etc...

Reply:
Ok. That was kinda the scenario I was picturing in my head.